src/Microsoft.AspNetCore.OData/Query/Query/DefaultSkipTokenHandler.cs

```diff
@@ -228,234 +300,154 @@ internal static string GenerateSkipTokenValue(Object lastMember, IEdmModel model
-        /// <returns></returns>
-        [SuppressMessage("Microsoft.Maintainability", "CA1506:AvoidExcessiveClassCoupling", Justification = "Class coupling acceptable.")]
-        private static IQueryable ApplyToCore(IQueryable query, ODataQuerySettings querySettings, IList<OrderByNode> orderByNodes, ODataQueryContext context, string skipTokenRawValue)
+        private static IQueryable ApplyToCore(IQueryable query, ODataQuerySettings querySettings,
+             ODataQueryOptions queryOptions, string skipTokenRawValue)
         {
-            Contract.Assert(query != null);
-            Contract.Assert(context.ElementClrType != null);
-
-            IDictionary<string, OrderByDirection> directionMap;
-            if (orderByNodes != null)
+            OrderByQueryOption orderByOption = queryOptions?.GenerateStableOrder();
+            if (orderByOption == null)
             {
-                directionMap =
-                    orderByNodes.OfType<OrderByPropertyNode>().ToDictionary(node => context.Model.GetClrPropertyName(node.Property), node => node.Direction);
-            }
-            else
-            {
-                directionMap = new Dictionary<string, OrderByDirection>();
+                return query;
             }
 
-            IDictionary<string, (object PropertyValue, Type PropertyType)> propertyValuePairs = PopulatePropertyValuePairs(skipTokenRawValue, context);
+            // It's better to visit the nodes of OrderByClause reclusively to get the orderby raw value.
+            // Since we don't have such methods, let's simply split the request raw value.
+            IList<string> orderBys = orderByOption.RawValue.Split(',');
+            IList<OrderByClause> orderByClauses = orderByOption.OrderByClause.ToList();
+            IList<string> tokenValueParis = PopulatePropertyValuePairs(skipTokenRawValue);
 
-            if (propertyValuePairs.Count == 0)
+            if (orderBys.Count != orderByClauses.Count && orderByClauses.Count != tokenValueParis.Count)
             {
                 throw Error.InvalidOperation(SRResources.SkipTokenProcessingError);
             }
 
-            bool parameterizeConstant = querySettings.EnableConstantParameterization;
-            ParameterExpression param = Expression.Parameter(context.ElementClrType);
-            Expression where = null;
-            /* We will create a where lambda of the following form -
-             * Where (Prop1>Value1)
-             * OR (Prop1=Value1 AND Prop2>Value2)
-             * OR (Prop1=Value1 AND Prop2=Value2 AND Prop3>Value3)
-             * and so on...
-             * Adding the first true to simplify implementation.
-             */
-            Expression lastEquality = null;
-            bool firstProperty = true;
-
-            foreach (KeyValuePair<string, (object PropertyValue, Type PropertyType)> item in propertyValuePairs)
+            string where = string.Empty;
+            string lastEquality = null;
```

Comment:
nit: The name `lastEquality` might give the impression that this refers to the last or final equality in the loop, but actually it refers to the equality generated in the previous iteration.
```suggestion
            string previousEquality = null;
```